qa/tasks/keycloak.py

```diff
@@ -0,0 +1,331 @@
+"""
+Deploy and configure Keycloak for Teuthology
+"""
+from io import BytesIO
+from io import StringIO
+from configobj import ConfigObj
+import base64
+import argparse
+import contextlib
+import logging
+import os
+import random
+import six
+import string
+import subprocess
+import json
+import sys
+from pathlib import Path
+
+from collections import OrderedDict
+from itertools import chain
+
+from teuthology import misc as teuthology
+from teuthology import contextutil
+from teuthology.config import config as teuth_config
+from teuthology.orchestra import run
+from teuthology.packaging import install_package
+from teuthology.packaging import remove_package
+from teuthology.exceptions import ConfigError
+
+log = logging.getLogger(__name__)
+
+def get_keycloak_dir(ctx):
+    return '{tdir}/keycloak-11.0.0'.format(tdir=teuthology.get_testdir(ctx))
+
+def run_in_keycloak_dir(ctx, client, args, **kwargs):
+    return ctx.cluster.only(client).run(
+        args=[ 'cd', get_keycloak_dir(ctx), run.Raw('&&'), ] + args,
+        **kwargs
+    )
+
+def get_toxvenv_dir(ctx):
+    return ctx.tox.venv_path
+
+def toxvenv_sh(ctx, remote, args, **kwargs):
+    activate = get_toxvenv_dir(ctx) + '/bin/activate'
+    return remote.sh(['source', activate, run.Raw('&&')] + args, **kwargs)
+
+@contextlib.contextmanager
+def install_packages(ctx, config):
+    """
+    Downloading the two required tar files
+    1. Keycloak
+    2. Wildfly (Application Server)
+    """
+    assert isinstance(config, dict)
+    log.info('Installing packages for Keycloak...')
+
+    for (client, _) in config.items():
+        (remote,) = ctx.cluster.only(client).remotes.keys()
+        test_dir=teuthology.get_testdir(ctx)
+        toxvenv_sh(ctx, remote, ['wget', 'https://downloads.jboss.org/keycloak/11.0.0/keycloak-11.0.0.tar.gz'])
+        toxvenv_sh(ctx, remote, ['tar', '-C', test_dir, '-xvzf', 'keycloak-11.0.0.tar.gz'])
+
+        tmp_dir=os.path.join(test_dir,"keycloak-11.0.0/")
+        
+        toxvenv_sh(ctx, remote, ['cd', tmp_dir, run.Raw('&&'), 'wget', 'https://downloads.jboss.org/keycloak/11.0.0/adapters/keycloak-oidc/keycloak-wildfly-adapter-dist-11.0.0.tar.gz'])
+        toxvenv_sh(ctx, remote, ['tar', '-C', tmp_dir, '-xvzf', '{tdr}/keycloak-wildfly-adapter-dist-11.0.0.tar.gz'.format(tdr=tmp_dir)])
+
+    try:
+        yield
+    finally:
+        log.info('Removing packaged dependencies of Keycloak...')
+        test_dir=teuthology.get_testdir(ctx)
+        tmp_dir=os.path.join(test_dir,"keycloak-11.0.0/")
```

Comment:
Here is another example of the keycloak version being hard coded that needs to be changed.